controller/array_action/ds8k_volume_cache.py

```diff
@@ -0,0 +1,66 @@
+from collections import defaultdict
+from threading import RLock
+
+from controller.common.csi_logger import get_stdout_logger
+
+logger = get_stdout_logger()
+
+
+class volumeCacheByAddress:
+    def __init__(self):
+        logger.debug("creating a new cache")
+        self.volume_cache_by_address = defaultdict(dict)
+        self.cache_lock = RLock()
+
+    def add(self, address, key, value):
+        logger.debug("adding {} to cache".format(key))
+        with self.cache_lock:
+            self.volume_cache_by_address[address][key] = value
+
+    def remove(self, address, key):
+        logger.debug("removing {} from cache".format(key))
+        with self.cache_lock:
+            if self.volume_cache_by_address[address].get(key) is not None:
+                del self.volume_cache_by_address[address][key]
+
+    def get(self, address, key):
+        logger.debug("getting {} from cache".format(key))
+        with self.cache_lock:
+            return self.volume_cache_by_address[address].get(key)
+
+    def get_all(self, address):
+        logger.debug("getting cache for address {}".format(address))
+        with self.cache_lock:
+            return self.volume_cache_by_address.get(address)

```

Comment:
if this function is only used in tests, then it should be implemented there instead.
and AFAIK the tests don't asynchronously share state, so this function probably won't need the lock either

Reply:
implement a class method that takes privet params out of the class?

Reply:
I think it would be best to just mock the cache, and assert the mock calls, without actually needing to compare entire dictionaries

Reply:
done